### ISSUE TITLE
load pre-trained weights

### ISSUE THREAD

**Issue by liu15509348793** [ultralytics/ultralytics#20273]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I want to load pre-trained weights for training, but my model has modified the detection head, so I want to load all weight information except the detection head for training. How can I choose to load some weights? Can you help me?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liu15509348793, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started. An Ultralytics engineer will assist you here soon 😊.

**Comment by Y-T-G** [comment]
You can use the `.load()` method.

```
model = YOLO("custom_yaml.yaml")
model.load("yolo11n.pt") 
```

**Comment by liu15509348793** [comment]
> You can use the `.load()` method.
> 
> ```
> model = YOLO("custom_yaml.yaml")
> model.load("yolo11n.pt") 
> ```

Thank you for your reply, but this loads all the pre-trained weights. I only want to load a certain part, that is, the parameters of other layers except the detection head. How should I do it?

**Comment by Y-T-G** [comment]
You can remove the head of the pretrained model and load that:

```python
>>> from ultralytics import YOLO
>>> model = YOLO("yolo11n.yaml")
>>> pt = YOLO("yolo11n.pt")
>>> del pt.model.model[-1]  # remove head
>>> pt.save("yolo11n_nohead.pt")  # save
>>> model.load("yolo11n_nohead.pt")  # load weights
Transferred 378/499 items from pretrained weights
```

**Comment by liu15509348793** [comment]
> You can remove the head of the pretrained model and load that:
> 
> >>> from ultralytics import YOLO
> >>> model = YOLO("yolo11n.yaml")
> >>> pt = YOLO("yolo11n.pt")
> >>> del pt.model.model[-1]  # remove head
> >>> pt.save("yolo11n_nohead.pt")  # save
> >>> model.load("yolo11n_nohead.pt")  # load weights
> Transferred 378/499 items from pretrained weights

OK, thank you very much, I will give it a try. Also, is there a more general way that allows me to arbitrarily choose to load or not load the parameters of a certain layer? Do I need to manually program and delete it?

**Comment by Y-T-G** [comment]
You will have to manually delete the weights if you don't want to load them

**Comment by liu15509348793** [comment]
> You will have to manually delete the weights if you don't want to load them

`model = YOLO('ultralytics/cfg/models/11/yolo11s-obb-new.yaml')
    pt = YOLO("train/yolo11s-obb/weights/best.pt")
    del pt.model.model[-1]  # remove head
    pt.save("yolo11s-obb/weights/best_nohead.pt")  # save
    `model.load('/yolo11s-obb/weights/best_nohead.pt')`
I modified the code above, and then：Transferred 437/439 items from pretrained weights
Freezing layer 'model.23.dfl.conv.weight'

**Comment by liu15509348793** [comment]
> You will have to manually delete the weights if you don't want to load them

model = YOLO('ultralytics/cfg/models/11/yolo11s-obb-new.yaml')
pt = YOLO("train/yolo11s-obb/weights/best.pt") 
del pt.model.model[-1]  # remove head 
pt.save("yolo11s-obb/weights/best_nohead.pt")  # save 
model.load('/yolo11s-obb/weights/best_nohead.pt')`
I modified the code above, and then：
Transferred 437/439 items from pretrained weights
Freezing layer 'model.23.dfl.conv.weight'
The number of layers loaded is the same as when I loaded it directly. Is this unsuccessful? My new model only changes the new detection head.

**Comment by Y-T-G** [comment]
By default, only matched weights are loaded. So if you modify a layer, the weights for those wouldn't be loaded because it doesn't match. So if your head is different then even if you don't remove the head, it will not be loaded. It will only load layers that match.

**Comment by liu15509348793** [comment]
The weights loaded by my modified new model are also from the new model. The number of layers loaded is consistent with the number of layers loaded after removing the detection head, so I feel that the detection head weights were not deleted correctly.For example, I give the following example
`model = YOLO('ultralytics/cfg/models/11/yolo11s-obb-new.yaml')
pt = YOLO("train/yolo11s-obb/weights/best.pt")
del pt.model.model[-1] # remove head
pt.save("yolo11s-obb/weights/best_nohead.pt") # save
model.load('/yolo11s-obb/weights/best_nohead.pt')``
and
`model = YOLO('ultralytics/cfg/models/11/yolo11s-obb-new.yaml')
model.load('/yolo11s-obb-new/weights/best.pt')``

The number of layers loaded in these two cases is the same

**Comment by Y-T-G** [comment]
You can print and check which layers are still present after deleting the last one 

`print(pt)`

**Comment by liu15509348793** [comment]
> You can print and check which layers are still present after deleting the last one
> 
> `print(pt)`

![Image](https://github.com/user-attachments/assets/47634db8-e3ce-4756-b19c-5b942d440842)

When I was training, I found that there were two places where the number of layers of migration was printed. I don’t know which one is correct? Why are there two printed?

**Comment by Y-T-G** [comment]
The first one is from your `.load()` code

**Comment by liu15509348793** [comment]
> The first one is from your `.load()` code


What does the second one show?

**Comment by Y-T-G** [comment]
Second one is from training. 

You can try passing the path to pretrained weights instead of `load()` by using `pretrained="yolo11_weights.pt"` inside `model.train()`.

**Comment by liu15509348793** [comment]
> pretrained


But I used load instead of pretrained, why is it printed a second time?

**Comment by glenn-jocher** [comment]
The training process is also attempting to load pretrained weights because when you call `.train()`, Ultralytics checks for weights to load regardless of whether you've already used `.load()`. 

If you want complete control over which weights get loaded, you can use `attempt_load_one_weight` from `ultralytics.nn.tasks` which gives you more direct access to the weight loading process:

```python
from ultralytics.nn.tasks import attempt_load_one_weight

# Load your custom model
model = YOLO('your_custom.yaml')

# Get the pretrained model but exclude the head
pretrained_model, _ = attempt_load_one_weight("path/to/weights.pt")
del pretrained_model.model[-1]  # Remove detection head

# Transfer weights manually
model.model.load(pretrained_model.state_dict())

# Start training without additional weight loading
model.train(pretrained=None, ...)
```

This approach gives you finer control over the weight loading process.
